bpf/lib/nodeport.h

```diff
@@ -1509,24 +1552,45 @@ int tail_nodeport_ipv4_dsr(struct __ctx_buff *ctx)
 	void *data, *data_end;
 	int ret, ohead = 0;
 	struct iphdr *ip4;
-	bool l2_hdr_required = true;
+	struct iphdr inner_ip4 __maybe_unused;
+	bool l2_hdr_required __maybe_unused = true;
+	__be32 node_addr __maybe_unused = 0;
+	__be32 backend_addr __maybe_unused = 0;
+	struct remote_endpoint_info *info __maybe_unused = NULL;
 
 	if (!revalidate_data(ctx, &data, &data_end, &ip4)) {
 		ret = DROP_INVALID;
 		goto drop_err;
 	}
 
-#if DSR_ENCAP_MODE == DSR_ENCAP_IPIP
+#if defined(ENABLE_DSR_TUNL)
+# if DSR_ENCAP_MODE == DSR_ENCAP_IPIP
+	backend_addr = ctx_load_meta(ctx, CB_ADDR_V4);
+	info = lookup_ip4_remote_endpoint(backend_addr);
+	if (!info) {
+		ret = DROP_INVALID;
+		goto drop_err;
+	}
+	node_addr = info->tunnel_endpoint;
+	ret = dsr_tunl_set_ipip4(ctx, ip4,
+					node_addr,
+					ctx_load_meta(ctx, CB_HINT), &ohead);
+# else
+#  error "Invalid load balancer DSR encapsulation mode!"
+# endif
+#else
```

Comment:
This is a bit difficult to read. One improvement would be:

```suggestion
#else /* ENABLE_DSR_TUNL */
```

But it might be better to have something like:

```
#if DSR_ENCAP_MODE == DSR_ENCAP_IPIP
#if defined(ENABLE_DSR_TUNL)
...
#else /* !ENABLE_DSR_TUNL */
...
#endif /* ENABLE_DSR_TUNL */
#else /* DSR_ENCAP_MODE */
...
#endif /* DSR_ENCAP_MODE */
```

Or maybe even:

```
#if DSR_ENCAP_MODE == DSR_ENCAP_IPIP
#if defined(TUNNEL_MODE)
...
#else /* !TUNNEL_MODE */
...
#endif /* TUNNEL_MODE */
#else /* DSR_ENCAP_MODE !=  DSR_ENCAP_IPIP */
...
#endif /* DSR_ENCAP_MODE */
```